src/ucs/datastruct/conn_match.c

```diff
@@ -0,0 +1,212 @@
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "conn_match.h"
+
+#include <ucs/debug/log.h>
+#include <ucs/debug/assert.h>
+#include <ucs/debug/memtrack.h>
+#include <ucs/algorithm/crc.h>
+
+
+static UCS_F_ALWAYS_INLINE uint32_t
+ucs_conn_match_addr_hash(ucs_conn_match_addr_t conn_match_addr)
+{
+    return ucs_crc32(0, (const void*)conn_match_addr.addr,
+                     conn_match_addr.length);
+}
+
+static UCS_F_ALWAYS_INLINE int
+ucs_conn_match_addr_equal(ucs_conn_match_addr_t conn_match_addr1,
+                          ucs_conn_match_addr_t conn_match_addr2)
+{
+    return (conn_match_addr1.length == conn_match_addr2.length) &&
+           !memcmp(conn_match_addr1.addr, conn_match_addr2.addr,
+                   conn_match_addr1.length);
+}
+
+KHASH_IMPL(ucs_conn_match, ucs_conn_match_addr_t, ucs_conn_match_entry_t, 1,
+           ucs_conn_match_addr_hash, ucs_conn_match_addr_equal);
+
+
+const char *ucs_conn_match_queue_title[] = {
+    [UCS_CONN_MATCH_QUEUE_EXP]   = "expected",
+    [UCS_CONN_MATCH_QUEUE_UNEXP] = "unexpected"
+};
+
+
+void ucs_conn_match_init(ucs_conn_match_ctx_t *conn_match_ctx,
+                         const ucs_conn_match_ops_t *ops)
+{
+    kh_init_inplace(ucs_conn_match, &conn_match_ctx->hash);
+    conn_match_ctx->ops = *ops;
+}
+
+void ucs_conn_match_cleanup(ucs_conn_match_ctx_t *conn_match_ctx)
+{
+    char dest_addr_str[UCS_CONN_MATCH_ADDR_STR_MAX];
+    ucs_conn_match_entry_t entry;
+    ucs_conn_match_addr_t conn_match_addr;
+
+    kh_foreach(&conn_match_ctx->hash, conn_match_addr, entry, {
+        if (!ucs_hlist_is_empty(&entry.conn_q[UCS_CONN_MATCH_QUEUE_EXP])) {
+            ucs_diag("match_ctx %p: addr %s expected queue is not empty",
+                     conn_match_ctx,
+                     conn_match_ctx->ops.addr_str(&conn_match_addr, dest_addr_str,
+                                                  UCS_CONN_MATCH_ADDR_STR_MAX));
+        }
+        if (!ucs_hlist_is_empty(&entry.conn_q[UCS_CONN_MATCH_QUEUE_UNEXP])) {
+            ucs_diag("match_ctx %p: addr %s unexpected queue is not empty",
+                     conn_match_ctx,
+                     conn_match_ctx->ops.addr_str(&conn_match_addr, dest_addr_str,
+                                                  UCS_CONN_MATCH_ADDR_STR_MAX));
+        }
+
+        ucs_free(conn_match_addr.addr);
+    })
+    kh_destroy_inplace(ucs_conn_match, &conn_match_ctx->hash);
+}
+
+static ucs_conn_match_entry_t*
+ucs_conn_match_entry_get(ucs_conn_match_ctx_t *conn_match_ctx,
+                         const ucs_conn_match_addr_t *dest_addr)
+{
+    char dest_addr_str[UCS_CONN_MATCH_ADDR_STR_MAX];
+    ucs_conn_match_entry_t *entry;
+    ucs_conn_match_addr_t *key;
+    khiter_t iter;
+    int ret;
+
+    iter  = kh_put(ucs_conn_match, &conn_match_ctx->hash,
+                   *dest_addr, &ret);
+    entry = &kh_value(&conn_match_ctx->hash, iter);
```

Comment:
looks like it is not safe to query value if `ret == UCS_KH_PUT_FAILED`
can you please check?

Reply:
fixed